Source/WebKit/Shared/PurgeableCachedDecodedImage.cpp

```diff
@@ -0,0 +1,153 @@
+public:
+    static PurgeableCachedDecodedImageTimer& singleton();
+
+    void start();
+    void defer();
+    void stop();
+
+private:
+    static void timerFired();
+
+    std::optional<DeferrableOneShotTimer> m_timer;
+};
+
+PurgeableCachedDecodedImageTimer& PurgeableCachedDecodedImageTimer::singleton()
+{
+    static LazyNeverDestroyed<PurgeableCachedDecodedImageTimer> timer;
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [] {
+        timer.construct();
+    });
+    return timer;
+}
+
+void PurgeableCachedDecodedImageTimer::start()
+{
+    if (!m_timer)
+        m_timer.emplace(&PurgeableCachedDecodedImageTimer::timerFired, cleanupDelay);
+
+    m_timer->restart();
+}
+
+void PurgeableCachedDecodedImageTimer::defer()
+{
+    if (m_timer)
+        m_timer->restart();
+}
+
+void PurgeableCachedDecodedImageTimer::stop()
+{
+    m_timer.reset();
+}
+
+void PurgeableCachedDecodedImageTimer::timerFired()
+{
+    WebProcess::singleton().releaseAllRemoteImageResources();
+    MemoryCache::singleton().destroyAndCacheDecodedDataForAllImages();
+}
+
+std::unique_ptr<PurgeableCachedDecodedImage> PurgeableCachedDecodedImage::create(ShareableBitmap& bitmap)
+{
+    auto memory = createVolatileCopyOfBitmap(bitmap);
+    if (!memory)
+        return nullptr;
+
+    return makeUnique<PurgeableCachedDecodedImage>(memory.releaseNonNull(), bitmap.size(), bitmap.configuration());
+}
+
+RefPtr<SharedMemory> PurgeableCachedDecodedImage::createVolatileCopyOfBitmap(ShareableBitmap& bitmap)
+{
+    auto memory = SharedMemory::allocate(bitmap.sizeInBytes());
+    if (!memory)
+        return nullptr;
+
+    memcpy(memory->data(), bitmap.data(), memory->size());
+    memory->setVolatile();
+    return memory;
+}
+
+PurgeableCachedDecodedImage::PurgeableCachedDecodedImage(Ref<SharedMemory>&& memory, const WebCore::IntSize& size, const ShareableBitmapConfiguration& configuration)
+    : m_memory(WTFMove(memory))
+    , m_size(size)
+    , m_configuration(configuration)
+{
+    PurgeableCachedDecodedImageTimer::singleton().defer();
+}
+
+PurgeableCachedDecodedImage::~PurgeableCachedDecodedImage() = default;
+
+RefPtr<NativeImage> PurgeableCachedDecodedImage::createNativeImage()
+{
+    if (!m_memory)
+        return nullptr;
+
+    if (m_memory->setNonVolatile() == WebCore::SetNonVolatileResult::Empty) {
+        m_memory = nullptr;
+        return nullptr;
+    }
+
+    auto bitmap = ShareableBitmap::create(m_size, m_configuration, m_memory.releaseNonNull());
+    if (!bitmap)
+        return nullptr;
+
+    auto decodedImage = CacheableDecodedImage::create(WTFMove(bitmap));
+    if (!decodedImage)
+        return nullptr;
+
+    return NativeImage::create(makeUniqueRefFromNonNullUniquePtr(WTFMove(decodedImage)));
+}
+
+void PurgeableCachedDecodedImage::startCleanup()
+{
+    PurgeableCachedDecodedImageTimer::singleton().start();
+}
+
+void PurgeableCachedDecodedImage::stopCleanup()
+{
+    PurgeableCachedDecodedImageTimer::singleton().stop();
+}
```

Comment:
Why do we need to include these static functions in this class? They release all the remote images not only a single one. So maybe these two functions can be static in `PurgeableCachedDecodedImageTimer` itself. But I think it should be renamed in this case since it will have the timer and the handlers as well.